Issue 227 create order wizard

dolchi21 opened the pull request:
Note: this is not the final product. I'm just creating this PR so that we can test and comment on it.
I made some changes to the flow. For example, asking fiatAmount before sats amount.
I have also made it obligatory to have a Community configured given that in the future the default community will cease to exist.


Fix #227

grunch commented:
I haven't tested but don't make that be obligatory to community, if the user have a default community use that, but if the user have `user.default_community_id === null` just create the order with `order.community_id = null`.

That way we can test it in production, communities are just starting 😄

grunch commented:
BTW I love how you started to modularize everything inside `bot/`, thank you!

dolchi21 commented:
What should we do about the help message and the tests? If user executes **/buy** alone the wizard will popup, not the help text.

> BTW I love how you started to modularize everything inside `bot/`, thank you!

Thanks! Hopefully it will result in easier codebase and less merge conflicts.

grunch commented:
I just tested, it works great, I just find a small issue, it allows to enter fiat_amount = 0 and amount = 0, then we have an error when creating the order, the easy way to fix this is we don't allow to people add fiat_amount=0, some people asked to allow to do the same we do with sats amount (allow 0) but if we do that someday, that should be a different issue, let's not complicate this PR.

About the message, I think if someone is using the wizard doesn't need to see the help response, we can tell to the user that wants to see the command that just entering `/sell h` the message will be displayed, but in fact that would work with anything you put there `/sell a`, `/buy b`, etc.

About the format of the code, could you please use prettier? I add a prettier configuration file which will help us to have coding standard and good practices.

dolchi21 commented:
> I just tested, it works great, I just find a small issue, it allows to enter fiat_amount = 0 and amount = 0, then we have an error when creating the order, the easy way to fix this is we don't allow to people add fiat_amount=0, some people asked to allow to do the same we do with sats amount (allow 0) but if we do that someday, that should be a different issue, let's not complicate this PR.

Updated the PR

> About the message, I think if someone is using the wizard doesn't need to see the help response, we can tell to the user that wants to see the command that just entering `/sell h` the message will be displayed, but in fact that would work with anything you put there `/sell a`, `/buy b`, etc.

Great. I modified the tests.

> About the format of the code, could you please use prettier? I add a prettier configuration file which will help us to have coding standard and good practices.

Done! I was only running **npm run lint**. Did not notice there was a **npm run format**

grunch commented:
We are almost there, I found some weird behavior, when I create a buy or sell and indicate the sats amount (not market price) the order is published in the correct channel but also the ad is sent to the user in the private chat with the bot.

grunch commented:
Testing without communities:

When you use the wizard without a community and ask you the currency it shows you a list but is not the complete list, maybe telegram adds limitation on the number of buttons that we can show, I think is a better idea when is not a community just ask for the currency, code.

dolchi21 commented:
> Other issue is `/community` command now is broken, I get this error
> 
> ```
> Error: Can't find scene: COMMUNITY_WIZARD_SCENE_ID
>     at SceneContextScene.enter (/home/grunch/dev/p2pbot/node_modules/telegraf/lib/scenes/context.js:54:19)
>     at /home/grunch/dev/p2pbot/bot/start.js:788:17
>     at processTicksAndRejections (node:internal/process/task_queues:96:5)
>     at async execute (/home/grunch/dev/p2pbot/node_modules/telegraf/lib/composer.js:468:17)
>     at async /home/grunch/dev/p2pbot/node_modules/telegraf/lib/composer.js:469:21
>     at async execute (/home/grunch/dev/p2pbot/node_modules/telegraf/lib/composer.js:468:17)
>     at async /home/grunch/dev/p2pbot/node_modules/telegraf/lib/composer.js:469:21
>     at async execute (/home/grunch/dev/p2pbot/node_modules/telegraf/lib/composer.js:468:17)
>     at async /home/grunch/dev/p2pbot/node_modules/telegraf/lib/composer.js:469:21
>     at async execute (/home/grunch/dev/p2pbot/node_modules/telegraf/lib/composer.js:468:17)
> ```

Apparently stage middleware in modules/orders/index was overwriting start.js stage middleware

grunch reviewed (changes requested):
bot object is not inside context and it's breaking creating orders using the old way, example

`/sell 600 500 ars xxx`

grunch commented on bot/modules/orders/commands.js:
```diff
@@ -0,0 +1,154 @@
+      await messages.currencyNotSupportedMessage(ctx, community.currencies);
+      return;
+    }
+    const order = await ordersActions.createOrder(ctx.i18n, ctx.bot, user, {
```
We need to pass bot object as a parameter of this function to use it here

dolchi21 replied:
ctx instead of ctx.bot should work as well. Ideally **createOrder** should not be sending messages. It should throw an exception and the caller should handle the error.

dolchi21 replied:
ctx instead of ctx.bot should work as well. Ideally **createOrder** should not be sending messages. It should throw an exception and the caller should handle the error.

grunch replied:
Yes we should handle errors in a better way, curious I didn't know that ctx == bot, it works

grunch commented:
ACK and merge, this PR is a piece of art 🤖
